src/checkout-buttons/strategies/masterpass-button-strategy.ts

```diff
@@ -0,0 +1,144 @@
+import { CheckoutButtonInitializeOptions, CheckoutButtonOptions } from '../';
+import { Checkout, CheckoutActionCreator, CheckoutStore } from '../../checkout';
+import {
+    InvalidArgumentError,
+    MissingDataError,
+    MissingDataErrorType
+} from '../../common/error/errors';
+import { bindDecorator as bind } from '../../common/utility';
+import { PaymentMethod, PaymentMethodActionCreator } from '../../payment';
+import {
+    Masterpass,
+    MasterpassCheckoutOptions,
+    MasterpassScriptLoader
+} from '../../payment/strategies/masterpass';
+
+import { CheckoutButtonStrategy, MasterpassButtonInitializeOptions } from './';
+
+export default class MasterpassButtonStrategy extends CheckoutButtonStrategy {
+    private _masterpassClient?: Masterpass;
+    private _signInButton?: HTMLElement;
+    private _paymentMethod?: PaymentMethod;
+    private _checkout?: Checkout;
+
+    constructor(
+        private _store: CheckoutStore,
+        private _checkoutActionCreator: CheckoutActionCreator,
+        private _paymentMethodActionCreator: PaymentMethodActionCreator,
+        private _masterpassScriptLoader: MasterpassScriptLoader
+    ) {
+        super();
+    }
+
+    initialize(options: CheckoutButtonInitializeOptions): Promise<void> {
+        const { masterpass: masterpassOptions, methodId } = options;
+
+        if (!masterpassOptions || !methodId) {
+            throw new InvalidArgumentError('Unable to proceed because "options.masterpass" argument is not provided.');
+        }
+
+        if (this._isInitialized) {
+            return super.initialize(options);
+        }
+
+        return Promise.all(
+            [
+                this._store.dispatch(this._paymentMethodActionCreator.loadPaymentMethod(methodId)),
+                this._store.dispatch(this._checkoutActionCreator.loadDefaultCheckout()),
+            ]
+        ).then(([statePayment, stateCheckout]) => {
+                this._paymentMethod = statePayment.paymentMethods.getPaymentMethod(methodId);
```

Comment:
You shouldn't have to retain `statePayment` and `stateCheckout`. Both are snapshots of the same state, but captured at different times; so because the checkout action is dispatched later, `stateCheckout` is fresher than `statePayment`. Anyway, my recommendation is to retrieve it directly from `store` so you know for sure you're dealing with the latest data.

```ts
const state = this._store.getState();

this._paymentMethod = state.paymentMethods.getPaymentMethod(methodId);
this._checkout = state.checkout.getCheckout();
```